### PR TITLE
[SC] Add IContract abstraction

### PR DESCRIPTION
**Pull request by rowandh**
Introduces IContract, which abstracts contract object creation and method invocation behind an interface.

### CODE REVIEW

File: src/Stratis.SmartContracts.Executor.Reflection/Contract.cs [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+        /// </summary>
+        private static void SetStateFields(SmartContract smartContract, ISmartContractState contractState)
+        {
+            FieldInfo[] fields = typeof(SmartContract).GetFields(BindingFlags.Instance | BindingFlags.NonPublic);
+
+            foreach (FieldInfo field in fields)
+            {
+                switch (field.Name)
+                {
+                    case "gasMeter":
```
Comment (codingupastorm) [inline src/Stratis.SmartContracts.Executor.Reflection/Contract.cs]: Would be nice to have some kinda nameof thing here but could be impossible.

Not blocking cos same in restorer at the moment.

File: src/Stratis.SmartContracts.Executor.Reflection/ReflectionVirtualMachine.cs [hunk 2]
```diff
@@ -67,61 +67,55 @@ public class ReflectionVirtualMachine : ISmartContractVirtualMachine
...
+                address,
+                transactionContext,
+                gasMeter,
+                repository,
+                internalTransferList);
 
-            var persistentState = new PersistentState(persistenceStrategy, contractAddress, this.network);
+            if (!contractLoadResult.IsSuccess)
+            {
+                // TODO this is temporary until we improve error handling overloads
```
[the remark this answers is not in the document]
Comment (rowandh) [reply src/Stratis.SmartContracts.Executor.Reflection/ReflectionVirtualMachine.cs]: We're creating a new Exception object as a container for the error message - it's not actually being thrown anywhere. In the future I'd like to just return any expected errors and only attach an exception if we caught one.

File: src/Stratis.SmartContracts.Executor.Reflection/ReflectionVirtualMachine.cs [hunk 3]
```diff
@@ -67,61 +67,55 @@ public class ReflectionVirtualMachine : ISmartContractVirtualMachine
...
+                var exception = new Exception(contractLoadResult.Error);
 
-            var internalTransferList = new List<TransferInfo>();
+                LogException(exception);
 
-            IInternalTransactionExecutor internalTransactionExecutor = this.internalTransactionExecutorFactory.Create(this, repository, internalTransferList, transactionContext);
+                this.logger.LogTrace("(-)[LOAD_CONTRACT_FAILED]:{0}={1}", nameof(gasMeter.GasConsumed), gasMeter.GasConsumed);
 
-            var balanceState = new BalanceState(repository, transactionContext.Amount, internalTransferList);
+                return VmExecutionResult.Error(gasMeter.GasConsumed, exception);
```
[the remark this answers is not in the document]
Comment (rowandh) [reply src/Stratis.SmartContracts.Executor.Reflection/ReflectionVirtualMachine.cs]: Exactly, future work 😃

File: src/Stratis.SmartContracts.Executor.Reflection/Contract.cs (hunk 1 above)
Comment (rowandh) [reply src/Stratis.SmartContracts.Executor.Reflection/Contract.cs]: Possible we could even just a single field to improve performance.

**Pull request merged**